src/main/java/frc/robot/subsystems/leds/LEDSubsystem.java

```diff
@@ -38,20 +50,116 @@ public void periodic() {
+        //if the color doesnt change, dont do anything
+        if(this.color.red == r && this.color.green == g && this.color.blue == b){ 
+            return;
+        }
         this.color = new Color(
-            (int) (red * BRIGHTNESS_SCALE_FACTOR),
-            (int) (green * BRIGHTNESS_SCALE_FACTOR),
-            (int) (blue * BRIGHTNESS_SCALE_FACTOR)
+            (int) (r * BRIGHTNESS_SCALE_FACTOR),
+            (int) (g * BRIGHTNESS_SCALE_FACTOR),
+            (int) (b * BRIGHTNESS_SCALE_FACTOR)
+        );
+        //sets the whole strip to the color
+        ledStrip.fillContinuousColor(this.color);
+        ledStrip.setContinuousColor();
+    }
+
+    /**
+     * Sets the driver-provided color of the LEDs from provided HSV values.
+     * @param h The hue component of the color, from [0, 180).
+     * @param s The saturation component of the color, from [0, 255].
+     * @param v The value component of the color, from [0, 255].
+     */
+    //sets the color to the 
+    public void setHSV(double h, double s, double v) {
+        this.color = Color.fromHSV(
+            (int) h,
+            (int) s,
+            (int) (v * BRIGHTNESS_SCALE_FACTOR)
         );
     }
+
+    public void toggleLEDControlMode(){
+        continuous = !continuous;
+    }
+
+    public void setManual(boolean manual){
+        this.manual = manual;
+    }
+
+    public void tagDetected(){
+        if(aprilTimer.hasElapsed(APRIL_BLINK_DURATION * 2)){
+            aprilTimer.reset();
+            aprilTimer.start();
+        }
+        aprilTimer2.reset();
+        aprilTimer2.start();
+    }
+
+    public void setColorPulse(){
+        Color currentColor;
+        //current color is the color that will be added at the bottom of the strip buffer
+        if(!aprilTimer.hasElapsed(APRIL_BLINK_DURATION)){
+            currentColor = APRIL_COLOR;
+        } else {
+            currentColor = color;
+        }
+        if(blinking){
+            ledStrip.fillContinuousColorIgnoringOneColor(BLINK_COLOR, APRIL_COLOR);
+        }
+        //if the leds are on update the continuous color and then set the leds to that continuous buffer
+        ledStrip.updateContinuousColor(currentColor);
+        ledStrip.setContinuousColor();
+    }
+
+    public void setTwoColor(){
+        Color color2;
+        if(blinking){
+            color = BLINK_COLOR;
+        }
+        if(!aprilTimer2.hasElapsed(.05)){
+            color2 = APRIL_COLOR;
+        } else {
+            color2 =  color;
+        }
+        ledStrip.setTwoColors(color, color2);
+    }
+
+    public void setColorFromXandY(double x, double y){
+        if(manual){
+            double angleRads = MathUtil.inputModulus(Math.atan2(y, x), 0, 2 * Math.PI);
+            this.color = Color.fromHSV(
+            (int) (angleRads / (2 * Math.PI) * 180),
+            (int) 255,
+            (int) (255 * BRIGHTNESS_SCALE_FACTOR)
+        );
+        } else {
+            if(x > INPUT_DEADZONE){
```

Comment:
Small thing but you can merge this nested if with the `else` into an `else if` to improve readability. Generally the more nested layers there are in a single method the harder it gets to trace code execution flow.